payload/to_value.go

```diff
@@ -164,35 +166,48 @@ func sliceToTFTupleValue(in []interface{}, st tftypes.Type, at *tftypes.Attribut
 
 func sliceToTFSetValue(in []interface{}, st tftypes.Type, at *tftypes.AttributePath) (tftypes.Value, error) {
 	il := make([]tftypes.Value, len(in), len(in))
-	oType := st.(tftypes.Set).ElementType
+	var oType tftypes.Type = tftypes.Type(nil)
 	for k, v := range in {
 		eap := at.WithElementKeyInt(int64(k))
 		iv, err := ToTFValue(v, st.(tftypes.Set).ElementType, at.WithElementKeyInt(int64(k)))
 		if err != nil {
 			return tftypes.Value{}, eap.NewErrorf("[%s] cannot convert list element [%d] to '%s': %s", eap, k, st.(tftypes.Set).ElementType.String(), err)
 		}
 		il[k] = iv
+		if oType == tftypes.Type(nil) {
+			oType = iv.Type()
+		}
 		if !oType.Is(iv.Type()) {
 			return tftypes.Value{}, eap.NewErrorf("[%s] conflicting list element type: %s", eap.String(), iv.Type())
 		}
 	}
+	// fallback for empty list, just use the requested type
+	if oType == tftypes.Type(nil) {
+		oType = st.(tftypes.Set).ElementType
+	}
 	return tftypes.NewValue(tftypes.Set{ElementType: oType}, il), nil
 }
 
 func mapToTFMapValue(in map[string]interface{}, st tftypes.Type, at *tftypes.AttributePath) (tftypes.Value, error) {
 	im := make(map[string]tftypes.Value)
-	oType := st.(tftypes.Map).AttributeType
+	var oType tftypes.Type
 	for k, v := range in {
 		eap := at.WithAttributeName(k)
 		mv, err := ToTFValue(v, st.(tftypes.Map).AttributeType, eap)
 		if err != nil {
 			return tftypes.Value{}, eap.NewErrorf("[%s] cannot convert map element '%s' to '%s': err", eap, st.(tftypes.Map).AttributeType.String(), err)
 		}
 		im[k] = mv
+		if oType == tftypes.Type(nil) {
+			oType = mv.Type()
+		}
 		if !oType.Is(im[k].Type()) {
 			return tftypes.Value{}, eap.NewErrorf("[%s] conflicting map element type: %s", eap.String(), mv.Type())
```

Comment:
On line 204, I got another one of those lint warnings for a potential nil pointer dereference. Just FYI. It could be fine.

Reply:
I think this time it's fine because line 202 initialises it if found to be nil.